### PR TITLE
feat: static variable analysis

### PR DESCRIPTION
**Pull request by jg-rp**
Statically analyze templates and report variable usage.

**Usage**

Retrieve the names of variables used in a template with `Liquid.variables(template)`. It returns an array of strings, one string for each distinct variable, without its properties.

```javascript
import { Liquid } from 'liquidjs'

const engine = new Liquid()

const template = engine.parse(`\
<p>
  {% assign title = user.title | capitalize %}
  {{ title }} {{ user.first_name | default: user.name }} {{ user.last_name }}
  {% if user.address %}
    {{ user.address.line1 }}
  {% else %}
    {{ user.email_addresses[0] }}
    {% for email in user.email_addresses %}
       - {{ email }}
    {% endfor %}
  {% endif %}
<p>
`)

console.log(engine.variablesSync(template))
```

**Output**

```javascript
[ 'user', 'title', 'email' ]
```

Alternatively, use `Liquid.fullVariables(template)` to get a list of variables including their properties. Notice that variables from tag and filter arguments are included too.

```javascript
// continued from above
engine.fullVariables(template).then(console.log)
```

**Output**

```javascript
[
  'user.title',
  'user.first_name',
  'user.name',
  'user.last_name',
  'user.address',
  'user.address.line1',
  'user.email_addresses[0]',
  'user.email_addresses',
  'title',
  'email'
]
```

Or use `Liquid.variableSegments(template)` to get an array of strings and numbers that make up each variable's path.

```javascript
// continued from above
engine.variableSegments(template).then(console.log)
```

**Output**

```javascript
[
  [ 'user', 'title' ],
  [ 'user', 'first_name' ],
  [ 'user', 'name' ],
  [ 'user', 'last_name' ],
  [ 'user', 'address' ],
  [ 'user', 'address', 'line1' ],
  [ 'user', 'email_addresses', 0 ],
  [ 'user', 'email_addresses' ],
  [ 'title' ],
  [ 'email' ]
]
```

### Global Variables

Notice, in the examples above, that `title` and `email` are included in the results. Often you'll want to exclude names that are in scope from `{% assign %}` tags, and temporary variables like those introduced by a `{% for %}` tag.

To get names that are expected to be _global_, that is, provided by application developers rather than template authors, use the `globalVariables`, `globalFullVariables` or `globalVariableSegments` methods (or their synchronous equivalents) of a `Liquid` class instance.

```javascript
// continued from above
engine.globalVariableSegments(template).then(console.log)
```

**Output**

```javascript
[
  [ 'user', 'title' ],
  [ 'user', 'first_name' ],
  [ 'user', 'name' ],
  [ 'user', 'last_name' ],
  [ 'user', 'address' ],
  [ 'user', 'address', 'line1' ],
  [ 'user', 'email_addresses', 0 ],
  [ 'user', 'email_addresses' ]
]
```

### Partial Templates

By default, LiquidJS will try to load and analyze any included and rendered templates too.

```javascript
import { Liquid } from 'liquidjs'

const footer = `\
<footer>
  <p>&copy; {{ "now" | date: "%Y" }} {{ site_name }}</p>
  <p>{{ site_description }}</p>
</footer>`

const engine = new Liquid({ templates: { footer } })

const template = engine.parse(`\
<body>
  <h1>Hi, {{ you | default: 'World' }}!</h1>
  {% assign some = 'thing' %}
  {% include 'footer' %}
</body>
`)

engine.globalVariables(template).then(console.log)
```

**Output**

```javascript
[ 'you', 'site_name', 'site_description' ]
```

You can disable analysis of partial templates by setting the `partials` options to `false`.

```javascript
// continue from above
engine.globalVariables(template, { partials: false }).then(console.log)
```

**Output**

```javascript
[ 'you' ]
```

If an `{% include %}` tag uses a dynamic template name (one that can't be determined without rendering the template) it will be ignored, even if `partials` is set to `true`.

### CODE REVIEW

File: src/tags/capture.ts
Comment (harttle) [inline src/tags/capture.ts:34]: This seems cleaner (less code) compared with previous implementation.

File: src/tags/cycle.ts
Comment (harttle) [inline src/tags/cycle.ts:51]: Would you consider return a iterator? Not sure how do you want to comsume the results. Can be less code but not as straight forward. Just an idea, your call.

```suggestion
  public *arguments (): IterableIterator<Argument> {
    yield *this.candidates.slice()
    if (this.group) yield this.group
  }
```

File: test/integration/static_analysis/variables.spec.ts
Comment (harttle) [inline test/integration/static_analysis/variables.spec.ts:104]: What if `d`, `a` are global while `b` is local? Not sure if we can implement it thoroughly. If not, would you consider to reduce this feature to a list of top level variables:

```
globals: ['d', 'a', 'b']
```

I guess this list is already useful for new comers, am I correct here?

File: src/tags/cycle.ts
Comment (jg-rp) [reply src/tags/cycle.ts]: That's a good idea. In my last commit I've tried accepting an `Iterable` from `children`, `arguments`, `blockScope` and `localScope`. Now tag implementations can choose to implement them as a generator or return an array, for example, whatever is more convenient.

**Comment by jg-rp** [comment]
With the latest commit, I've changed several of the built-in tags to fix their row and column numbers reported from `Token.getPosition()`.

I guess this is a good point to decide if having correct row and column is important enough to warrant these changes.

Notice that when parsing `if`, `elsif` and `unless` tags, we're now avoiding some string slicing. Instead, when `new Value()` is given a `TagToken`, we pass the entire input string and a range to to `Tokenizer`, which I'm hoping will result in a performance boost rather than a performance penalty.

File: src/template/value.ts
Comment (harttle) [reply src/template/value.ts:17]: Update: sorry I just find `tagToken.tokenizer` maybe already the tokenizer we want to expose with `argsTokenizer`, could you try to use it instead of constructing a new one?

File: test/integration/util/error.spec.ts
Comment (harttle) [inline test/integration/util/error.spec.ts:267]: Great job!

File: src/template/value.ts
Comment (jg-rp) [reply src/template/value.ts]: 👍 That's much better.

File: test/integration/static_analysis/variables.spec.ts
Comment (jg-rp) [reply test/integration/static_analysis/variables.spec.ts]: I've added a test with a local variable nested within a global variable. Still not sure what's best here.

**Comment by jg-rp** [comment]
I've provisionally implemented some convenience analysis methods on the `Liquid` class. Please consider these (along with any other features) as ideas that can be removed before merging.

Other ideas that I've yet to implement:

- Report names of Liquid filters and their locations in analysis results (I've seen people ask for this before).
- Report names of tags and their locations in analysis results.
- Add options to control partial template analysis. At the moment we throw an error if a partial template can't be loaded, and silently ignore partial templates included/rendered with a dynamic name.

**Comment by jg-rp** [comment]
@harttle, I've not forgotten about this or you previous comments about keeping track of aliases. I'll get back to it soon.

**Review by harttle — APPROVED** [review]

File: src/liquid.ts
Comment (jg-rp) [reply src/liquid.ts:130]: These async functions are here for the benefit of  `include`, `render` and `layout` tags, which all call `liquid._parsePartialFile()` in their `children()` method if the `partials` option is set to `true` (the default).

Supporting automatic loading and analysis of partial templates does add a lot of noise and complicates the `Template` interface. I'll happily remove this and simplify things if you think this feature is not worth including.

File: docs/source/tutorials/static-analysis.md
Comment (jg-rp) [reply docs/source/tutorials/static-analysis.md:71]: Using your example, `a[b.c].d`, `variableSegments` was incorrectly producing something like this:

```js
[
  [
    'a',
    Variable {
      segments: [ 'b', 'c' ],
      location: { row: 1, col: 6, file: undefined }
    },
    'd'
  ],
  [ 'b', 'c' ]
]
```

This was not my intention. Now we get the following.

**variables**
```javascript
[ 'a', 'b' ]
```

**fullVariables**
```javascript
[
  'a[b.c].d',
  'b.c'
]`
```

**variableSegments**

```javascript
[
  [ 'a', [ 'b', 'c' ], 'd' ],
  [ 'b', 'c' ]
]
```

For `arr[0].length` and `arr.length`,  we get `[ 'arr', 0, 'length' ]` and `[ 'arr', 'length' ]`, respectively.

Is that what you had in mind?

**Comment by jg-rp** [comment]
Can you think of better names for `fullVariables`, `globals`, `locals`, and/or `segments`?

At the moment:

- The name "global" or "global variables" is inspired by Python's built-in `globals()` function. In our case it means names added to a template's scope by application developers at render time. Global variables are available to the _root_ template and any templates that it includes, including those rendered with the `{% render %}` tag.

- The name "locals" or "template local variables" is inspired by Python's built-in `locals()` function. Here it means names that have been added to a template's scope from an `{% assign %}`, `{% capture %}`, `{% increment %}`, etc. tag.

- "Segments" is a term used by [RFC 9535](https://datatracker.ietf.org/doc/html/rfc9535). One can think of Liquid variables as _paths_, where each path is made up of one or more segments.

- A "fullVariable" is roughly equivalent to a "normalized path" in [RFC 9535](https://datatracker.ietf.org/doc/html/rfc9535).

File: docs/source/tutorials/static-analysis.md
Comment (harttle) [inline docs/source/tutorials/static-analysis.md:1]: I think you also need change sidebar and em.yml, to make this file visible on sidebar.

File: docs/source/tutorials/static-analysis.md
Comment (harttle) [reply docs/source/tutorials/static-analysis.md:71]: No, I mean maybe we can drop nesting. And treat them as different references. I guess what exactly is inside ’[]’ is not important, we only know that array/map is accessed. Assuming figure out its value when called is not feasible for static analysis, then ppl won’t use that information effectively. Here’s my simplified approach 

variables

[ 'a', 'b' ]
fullVariables

[
  'a[].d',
  'b.c'
]`
variableSegments

[
  [ 'a', [ 'd' ] ],
  [ 'b', 'c' ]
]

note the last one use nested array to represent entering into array, not nested index.

**Comment by harttle** [comment]
your current naming is OK for me. Sorry some comments are not submitted yesterday, causing confusion.

File: docs/source/tutorials/static-analysis.md
Comment (jg-rp) [reply docs/source/tutorials/static-analysis.md:71]: Ah, I see. I think your approach should be in addition to `variableSegments`, as separate method/s. Then users have the option of working with nested variables and array indexes, if they're useful, or your more convenient representation if they're not.

Perhaps `normalizedSegments` would be a good name?

It seems like we're loosing potentially valuable information if we drop nested variables altogether.

Comment (harttle) [reply docs/source/tutorials/static-analysis.md:71]: Just another idea to simplify the implementation. If you still think dynamic index info is important, we can keep your current implementation. No need to compromise for my opinion.

**Comment by harttle** [comment]
I can help update the docs afterwards. Are we good to merge this PR now?

**Comment by jg-rp** [comment]
> I can help update the docs afterwards. Are we good to merge this PR now?

There's still a couple of corner cases to be addressed when tracking aliased variables in deeply nested `render`ed templates and handling Jekyll-style includes. Otherwise, if you're happy, it's good to go.

**Comment by harttle** [comment]
Merged and ddded an experimental notice in documents so we people know we're improving it.

**Pull request merged**
